mitxgraders/comparers/comparers.py

```diff
@@ -178,3 +180,131 @@ def eigenvector_comparer(comparer_params_evals, student_eval, utils):
+    False
+
+    Complex scale factors work, too:
+    >>> grader(None, '(4 + 2*i)*[3, x, 1 + i]')['ok']
+    True
+
+    Student input should be nonzero:
+    >>> result = grader(None, '[0, 0, 0]')
+    >>> result['ok']
+    False
+    >>> result['msg']
+    'Input should be a nonzero vector.'
+
+    Input shape is validated:
+    >>> grader(None, '5')
+    Traceback (most recent call last):
+    InputTypeError: Expected answer to be a vector, but input is a scalar
+
+    Multiple vectors can be provided:
+    >>> grader = MatrixGrader(
+    ...     answers={
+    ...         'comparer_params': [
+    ...             '[1, 1, 0]',    # v0
+    ...             '[0, 1, 2]'     # v1
+    ...         ],
+    ...         'comparer': vector_span_comparer
+    ...     },
+    ... )
+
+    The vector 2*v0 + 3i*v1 = [2, 2+3i, 6i] is in the span of v0 and v1:
+    >>> grader(None, '[2, 2 + 3*i, 6*i]')['ok']
+    True
+    """
+
+    # All comparer_param_evals should have same shape, or numpy.linalg.lstsq
+    # will let us know.
+    utils.validate_shape(student_eval, comparer_params_evals[0].shape)
+
+    if utils.within_tolerance(0, np.linalg.norm(student_eval)):
+        return {
+            'ok': False,
+            'grade_decimal': 0,
+            'msg': 'Input should be a nonzero vector.'
+        }
+
+    # Use ordinary least squares to find an approximation to student_eval
+    # that lies within the span of given vectors, then check that the
+    # residual-sum is small in comparison to student input.
+    column_vectors = np.array(comparer_params_evals).transpose()
+    ols = np.linalg.lstsq(column_vectors, student_eval)
+    error = np.sqrt(ols[1])
+
+    # Check that error is nearly zero, using student_eval as a reference
+    # when tolerance is specified as a percentage
+    return is_nearly_zero(error, student_eval, utils.tolerance)
+
+def vector_phase_comparer(comparer_params_evals, student_eval, utils):
+    """
+    Check that student input equals a given input (to within tolerance), up to
+    an overall phase factor.
+
+    comparer_params: [target_vector]
+
+    Usage
+    =====
+
+    >>> from mitxgraders import MatrixGrader
+    >>> grader = MatrixGrader(
+    ...     answers={
+    ...         'comparer_params': [
+    ...             '[1, exp(-i*phi)]',
+    ...         ],
+    ...         'comparer': vector_phase_comparer
+    ...     },
+    ...     variables=['phi'],
+    ... )
+
+    >>> grader(None, '[1, exp(-i*phi)]')['ok']
+    True
+    >>> grader(None, '[exp(i*phi/2), exp(-i*phi/2)]')['ok']
+    True
+    >>> grader(None, '[i, exp(i*(pi/2 - phi))]')['ok']
+    True
+
+    >>> grader(None, '[1, exp(+i*phi)]')['ok']
+    False
+    >>> grader(None, '[2, 2*exp(-i*phi)]')['ok']
+    False
+
+    """
+    # We'll check that student input is in the span as target vector and that
+    # it has the same magnitude
+
+    in_span = vector_span_comparer(comparer_params_evals, student_eval, utils)
```

Comment:
Can we add a check here that `comparer_params_evals` has a single entry and is a vector?

Reply:
I've added this sort of validation for `vector_span_comparer` and `vector_phase_comparer`. Two downsides:

1. Will run for each sample during grading
2. Raises `StudentFacingError`, since validation happens upon "check" not upon instantiation.

Reply:
(Minor downsides, not worth worrying about right now. Overcoming them would require validating comparer_params during instantiation, which would be a lot of refactoring. We have basically the same issue for validating author answers with default comparer.)

Reply:
I agree - given that we don't validate author answers until check (and to be fair, neither does edX!), these are indeed minor downsides.